src/NUnitFramework/framework/Internal/PreFilter.cs

```diff
@@ -0,0 +1,208 @@
+
+        /// <summary>
+        /// Return a new PreFilter, without elements, which is considered
+        /// empty and always matches.
+        /// </summary>
+        public static PreFilter Empty
+        {
+            get { return new PreFilter(); }
+        }
+
+        /// <summary>
+        /// Return true if the filer is empty, in which case it
+        /// always succeeds. Technically, this is just a filter and
+        /// you can add elements but it's best to use Empty when
+        /// you need an empty filter and new when you plan to add.
+        /// </summary>
+        public bool IsEmpty
+        {
+            get { return _filters.Count == 0; }
+        }
+
+        /// <summary>
+        /// Add a new filter element to the filter
+        /// </summary>
+        /// <param name="filterText"></param>
+        public void Add(string filterText)
+        {
+            _filters.Add(new FilterElement(filterText));
+        }
+
+        /// <summary>
+        /// Use the filter on a Type, returning true if the type matches the filter
+        /// and should therefore be included in the discovery process.
+        /// </summary>
+        public bool Match(Type type)
+        {
+            if (IsEmpty)
+                return true;
+
+            string typeName = type.FullName;
+
+            foreach (FilterElement filter in _filters)
+            {
+                if (filter.Match(type))
+                    return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Use the filter on a Type, returning true if the type matches the filter
+        /// and should therefore be included in the discovery process.
+        /// </summary>
+        public bool Match(Type type, MethodInfo method)
+        {
+            if (IsEmpty)
+                return true;
+
+            int filterCount = 0;
+            foreach (FilterElement filter in _filters)
+            {
+                if (type.FullName == filter.ClassName)
+                {
+                    ++filterCount;
+                    if (method.Name == filter.MethodName)
+                        return true;
+                }
+            }
+
+            // If no match, we succeed only if there were no method filters for this fixture
+            return filterCount == 0;
+        }
+
+        #region Nested FilterElementType Enumeration
+
+        private enum FilterElementType
+        {
+            Unknown,
+            Namespace,
+            Fixture,
+            Method
+        }
+
+        #endregion
+
+        #region Nested FilterElement Class
+
+        private class FilterElement
+        {
+            public FilterElementType ElementType = FilterElementType.Unknown;
+            public string Text;
+            public string ClassName;
+            public string MethodName;
+
+            public FilterElement(string text)
+            {
+                int argStart = text.IndexOfAny(ARG_START);
```

Comment:
`ARG_START` looks for `<`, but `MatchFixtureElement` only works with the ECMA `` `1[`` format since you're comparing Type.FullName. Which format do we expect in existing places? Also, what about nested classes- do we expect users to type `+` or `.` for the separator?

Reply:
Good point! Note that the user may type it in as a prefilter or as a test (which gets converted to a prefilter), so it could be either `<` or `[`. For nested classes, I think we're in luck because we use + in all cases. Finally, I would never expect the user to type in a backslash.

If you have any ideas for test cases, I'll add them.